### PR TITLE
liquidity

### PR DESCRIPTION
**Pull request by talbaneth**

### CODE REVIEW

File: contracts/LiquidityConversionRates.sol [hunk 1]
```diff
@@ -3,55 +3,136 @@ pragma solidity ^0.4.18;
 import "./ConversionRates.sol";
 import "./LiquidityFormula.sol";
 
-contract LiquidityConversionRates is ConversionRates, LiquidityFormula {
+contract LiquidityConversionRates is ConversionRatesInterface, LiquidityFormula, Withdrawable, Utils {
+    ERC20 token;
     uint public r;
     uint public Pmin;
-    uint public precision;
+    uint public qtyPrecision;
```
Comment (yaronvel) [inline contracts/LiquidityConversionRates.sol]: it is not exactly qty, as r and Pmin are not qty. Please try to think of better name. Maybe `formulaPrecision`?

File: contracts/LiquidityConversionRates.sol [hunk 2]
```diff
@@ -3,55 +3,136 @@ pragma solidity ^0.4.18;
...
     uint public numPrecisionBits;
```
Comment (yaronvel) [inline contracts/LiquidityConversionRates.sol]: `numQtyPrecisionBits`?

File: contracts/LiquidityConversionRates.sol [hunk 3]
```diff
@@ -3,55 +3,136 @@ pragma solidity ^0.4.18;
...
+    uint public maxEthCapBuy;
```
Comment (yaronvel) [inline contracts/LiquidityConversionRates.sol]: `InWeiResultion`? `InQtyPrecisionResolution`? Please make it part of the name.

File: contracts/LiquidityConversionRates.sol [hunk 4]
```diff
@@ -3,55 +3,136 @@ pragma solidity ^0.4.18;
...
+    uint public maxEthCapSell;
+    uint public collectedTokenFees;
```
Comment (yaronvel) [inline contracts/LiquidityConversionRates.sol]: `InWei`?

File: contracts/LiquidityConversionRates.sol [hunk 5]
```diff
@@ -3,55 +3,136 @@ pragma solidity ^0.4.18;
...
 
+    function LiquidityConversionRates(address _admin) public{
+        admin = _admin;
+    }
 
-    function LiquidityConversionRates(address _admin) public ConversionRates(_admin)
-        { } // solhint-disable-line no-empty-blocks
-
-    function setLiquidityParams(uint _r, uint _Pmin,uint _numPrecisionBits) public onlyAdmin {
+    function setLiquidityParams(ERC20 _token, uint _r, uint _Pmin, uint _numPrecisionBits, uint _maxEthCapBuy, uint _maxEthCapSell, uint _feeInBps) public onlyAdmin {
```
Comment (yaronvel) [inline contracts/LiquidityConversionRates.sol]: please add resolution for every variable name.

File: contracts/LiquidityConversionRates.sol [hunk 6]
```diff
@@ -3,55 +3,136 @@ pragma solidity ^0.4.18;
...
-    uint public precision;
+    uint public qtyPrecision;
     uint public numPrecisionBits;
+    uint public maxEthCapBuy;
+    uint public maxEthCapSell;
+    uint public collectedTokenFees;
+    uint public feeInBps;
 
+    function LiquidityConversionRates(address _admin) public{
+        admin = _admin;
```
Comment (yaronvel) [inline contracts/LiquidityConversionRates.sol]: `transferAdminQuickly(_admin)` is more elegant.

File: contracts/LiquidityConversionRates.sol [hunk 7]
```diff
@@ -3,55 +3,136 @@ pragma solidity ^0.4.18;
...
         if(buy) {
           // ETH goes in, token goes out
-          deltaE = qty * precision / 10**ETH_DECIMALS;
-          if(deltaE == 0) return 0;
+          deltaE = fromWeiToQty(qty);
+          if(deltaE == 0) deltaE = 1;
 
-          deltaT = deltaTFunc(r,Pmin,E,deltaE,precision);
+          deltaT = deltaTFunc(r,Pmin,E,deltaE,qtyPrecision);
+          deltaT = reduceFee(deltaT);
```
Comment (yaronvel) [inline contracts/LiquidityConversionRates.sol]: technically, you can call it once, on the rate.

File: contracts/LiquidityConversionRates.sol [hunk 8]
```diff
@@ -3,55 +3,136 @@ pragma solidity ^0.4.18;
...
     }
 
-    function getRate(ERC20 token, uint currentBlockNumber, bool buy, uint qty) public view returns(uint) {
-        uint E = reserveContract.balance * precision / 10**ETH_DECIMALS;
+    function getRate(ERC20 conversionToken, uint currentBlockNumber, bool buy, uint qty) public view returns(uint) {
+
+        currentBlockNumber;
+
+        uint E = fromWeiToQty(conversionToken.balance);
+        // Yaron - here we need to reduce what was calculated for fee? I think not since we deal with token fee now.
```
Comment (yaronvel) [inline contracts/LiquidityConversionRates.sol]: we don't

File: contracts/LiquidityConversionRates.sol [hunk 9]
```diff
@@ -3,55 +3,136 @@ pragma solidity ^0.4.18;
...
+    {
+        conversionToken;
+        rateUpdateBlock;
+        currentBlock;
+
+        uint buyQty = fromDecimalsToQty(abs(buyAmount));
+        uint feeQty = getFee(buyQty);
+        collectedTokenFees += feeQty;
+
+        // Yaron - do we need? -require(msg.sender == reserveContract);
```
Comment (yaronvel) [inline contracts/LiquidityConversionRates.sol]: yes

File: contracts/LiquidityConversionRates.sol [hunk 10]
```diff
@@ -3,55 +3,136 @@ pragma solidity ^0.4.18;
...
+        // Yaron - do we need? - if (rateUpdateBlock == 0) rateUpdateBlock = getRateUpdateBlock(token);
```
Comment (yaronvel) [inline contracts/LiquidityConversionRates.sol]: no

File: contracts/LiquidityConversionRates.sol [hunk 11]
```diff
@@ -3,55 +3,136 @@ pragma solidity ^0.4.18;
...
+        // Yaron - do we need it - return addImbalance(token, buyAmount, rateUpdateBlock, currentBlock);
```
Comment (yaronvel) [inline contracts/LiquidityConversionRates.sol]: no

File: contracts/LiquidityConversionRates.sol [hunk 12]
```diff
@@ -3,55 +3,136 @@ pragma solidity ^0.4.18;
...
+    }
+
+    function resetCollectedFees() public {
```
Comment (yaronvel) [inline contracts/LiquidityConversionRates.sol]: only admin!!!!!!!!!!!!!!!!!!!!!!!

File: contracts/LiquidityConversionRates.sol [hunk 13]
```diff
@@ -3,55 +3,136 @@ pragma solidity ^0.4.18;
...
+        collectedTokenFees = 0;
+    }
+
+    function getCollectedFees() public view returns(uint) { //Yaron - Do we need it?
```
Comment (yaronvel) [inline contracts/LiquidityConversionRates.sol]: if you want to have it, call it `getCollectedFeesInTokenWei`.

File: contracts/LiquidityConversionRates.sol [hunk 14]
```diff
@@ -3,55 +3,136 @@ pragma solidity ^0.4.18;
...
+        return fromQtyToWei(collectedTokenFees);
+    }
+
+    function fromDecimalsToQty(uint amount) internal view returns(uint) {
```
Comment (yaronvel) [inline contracts/LiquidityConversionRates.sol]: The standard term is `tokenWei` or `Twei`.

File: contracts/LiquidityConversionRates.sol [hunk 15]
```diff
@@ -3,55 +3,136 @@ pragma solidity ^0.4.18;
...
+    function fromQtyToWei(uint qty) internal view returns(uint) {
+        return (qty * 10**ETH_DECIMALS) / qtyPrecision;
+    }
+
+    function reduceFee(uint val) internal view returns(uint) {
+        return ((10000 - feeInBps) * val) / 10000;
+    }
 
-        return getRateWithE(token,currentBlockNumber,buy,qty,E);
+    function getFee(uint val) internal view returns(uint) {
```
Comment (yaronvel) [inline contracts/LiquidityConversionRates.sol]: calculation is wrong. Also name is bad.

File: contracts/LiquidityConversionRates.sol (hunk 12 above)
Comment (yaronvel) [reply contracts/LiquidityConversionRates.sol]: here event is needed

File: contracts/LiquidityConversionRates.sol [hunk 16]
```diff
@@ -3,55 +3,136 @@ pragma solidity ^0.4.18;
...
           Pmin = _Pmin;
-          precision = uint(1)<<_numPrecisionBits;
+          qtyPrecision = uint(1)<<_numPrecisionBits;
           numPrecisionBits = _numPrecisionBits;
+          maxEthCapBuy = fromWeiToQty(_maxEthCapBuy);
+          maxEthCapSell = fromWeiToQty(_maxEthCapSell);
+          collectedTokenFees = 0;
+          require(_feeInBps < 10000);
+          feeInBps = _feeInBps;
     }
```
Comment (yaronvel) [inline contracts/LiquidityConversionRates.sol]: event would be nice here

File: contracts/LiquidityConversionRates.sol (hunk 1 above)
Comment (talbaneth) [reply contracts/LiquidityConversionRates.sol]: using formulaPrecision..

File: contracts/LiquidityConversionRates.sol (hunk 2 above)
Comment (talbaneth) [reply contracts/LiquidityConversionRates.sol]: numFpBits

File: contracts/LiquidityConversionRates.sol (hunk 3 above)
Comment (talbaneth) [reply contracts/LiquidityConversionRates.sol]: maxCapBuyInFp.

File: contracts/LiquidityConversionRates.sol (hunk 4 above)
Comment (talbaneth) [reply contracts/LiquidityConversionRates.sol]: collectedFeesInTwei.

File: contracts/LiquidityConversionRates.sol (hunk 6 above)
Comment (talbaneth) [reply contracts/LiquidityConversionRates.sol]: done

File: contracts/LiquidityConversionRates.sol (hunk 8 above)
Comment (talbaneth) [reply contracts/LiquidityConversionRates.sol]: ok

File: contracts/LiquidityConversionRates.sol (hunk 9 above)
Comment (talbaneth) [reply contracts/LiquidityConversionRates.sol]: done.
added a setReserveAddress function with admin permission.

File: contracts/LiquidityConversionRates.sol (hunk 10 above)
Comment (talbaneth) [reply contracts/LiquidityConversionRates.sol]: ok

File: contracts/LiquidityConversionRates.sol (hunk 11 above)
Comment (talbaneth) [reply contracts/LiquidityConversionRates.sol]: ok

File: contracts/LiquidityConversionRates.sol (hunk 12 above)
Comment (talbaneth) [reply contracts/LiquidityConversionRates.sol]: done onlyAdmin.
added an event.

File: contracts/LiquidityConversionRates.sol (hunk 13 above)
Comment (talbaneth) [reply contracts/LiquidityConversionRates.sol]: removed it, now wesave in storage collectedFeesInTwei, so no additional value for a dedicated function.

File: contracts/LiquidityConversionRates.sol (hunk 14 above)
Comment (talbaneth) [reply contracts/LiquidityConversionRates.sol]: using FromTweiToFp() now.

File: contracts/LiquidityConversionRates.sol (hunk 16 above)
Comment (talbaneth) [reply contracts/LiquidityConversionRates.sol]: added an event.

File: contracts/LiquidityConversionRates.sol [hunk 17]
```diff
@@ -1,57 +1,249 @@
...
-    uint public precision;
-    uint public numPrecisionBits;
 
+contract LiquidityConversionRates is ConversionRatesInterface, LiquidityFormula, Withdrawable, Utils {
+    ERC20 public token;
+    uint public rInFp;
+    uint public pMinInFp;
+    uint public formulaPrecision;
+    uint public numFpBits;
+    uint public maxCapBuyInFp;
```
Comment (yaronvel) [inline contracts/LiquidityConversionRates.sol]: empty lines between some of the variables would help...
e.g., `maxCapBuyInFp` and `maxCapSellInFp` can be together.

File: contracts/LiquidityConversionRates.sol [hunk 18]
```diff
@@ -1,57 +1,249 @@
...
+    uint public maxCapSellInFp;
```
Comment (yaronvel) [inline contracts/LiquidityConversionRates.sol]: this is in ether. maybe change the name to reflect it? also the buy thing.

File: contracts/LiquidityConversionRates.sol [hunk 19]
```diff
@@ -1,57 +1,249 @@
...
+    uint public collectedFeesInTwei;
+    uint public feeInBps;
+    uint public maxBuyRateInPRECISION;
```
Comment (yaronvel) [inline contracts/LiquidityConversionRates.sol]: why `PRECISION` and not `Precision`?

File: contracts/LiquidityConversionRates.sol [hunk 20]
```diff
@@ -1,57 +1,249 @@
...
-          if(deltaT == 0) return 0;
 
-          deltaE = deltaEFunc(r,Pmin,E,deltaT,precision,numPrecisionBits);
-          rate = deltaE * PRECISION / deltaT;
+        rateInPRECISION = validateRate(rateInPRECISION, buy);
+        require(rateInPRECISION < MAX_RATE);
+        return rateInPRECISION;
+    }
+
+    function validateRate(uint rateInPRECISION, bool buy) public view returns(uint) {
```
Comment (yaronvel) [inline contracts/LiquidityConversionRates.sol]: name of function is bad.
It is not boolean validation.
Please find better name.

**Pull request merged**